### PR TITLE
Large steps support & shape optimization tutorial

### PR DESCRIPTION
**Pull request by bathal1**
## Description

This PR brings the following changes:
- Added support for the algorithm from the "[Large Steps in Inverse Rendering of Geometry](http://rgl.epfl.ch/publications/Nicolet2021Large)" paper in Mitsuba. It currently only support meshes.
- Added some additional to `Mesh::parameters_changed` to avoid unwanted behavior when updating a given mesh's tesselation, e.g. through remeshing
- Added a "Shape optimization" tutorial, that ewplains how to use the "Large Steps" method for shape optimization.

This would conclude the discussion in #600.

### CODE REVIEW

File: src/python/python/ad/largesteps.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    return idx, values
+
+class SolveCholesky(dr.CustomOp):
+    """
+    DrJIT custom operator to solve a linear system using a Cholesky factorization.
+    """
+
+    def eval(self, solver, u):
+        self.solver = solver
+        x = dr.empty(dr.cuda.TensorXf, shape=u.shape)
```
Comment (DoeringChristian) [inline src/python/python/ad/largesteps.py:36]: Changing the type to `mi.TensorXf` allows this to work in llvm modes as well i.e.
```python
        x = dr.empty(mi.TensorXf, shape=u.shape)
```

File: src/python/python/ad/largesteps.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        solver.solve(u, x)
+        return mi.TensorXf(x)
+
+    def forward(self):
+        x = dr.empty(mi.TensorXf, shape=self.grad_in('u').shape)
+        self.solver.solve(self.grad_in('u'), x)
+        self.set_grad_out(x)
+
+    def backward(self):
+        x = dr.empty(dr.cuda.TensorXf, shape=self.grad_out().shape)
```
Comment (DoeringChristian) [inline src/python/python/ad/largesteps.py:46]: Changing the type to `mi.TensorXf` allows this to work in llvm modes as well i.e.
```python
        x = dr.empty(mi.TensorXf, shape=self.grad_out().shape)

```

**Comment by DoeringChristian** [comment]
Hi @bathal1,

Thanks for your implementation it is definitively the way to do it properly :sweat_smile:.
I have tested your implementation as there where some problems with my implementation using the `pytorch` interface (see [here](https://github.com/mitsuba-renderer/mitsuba3/discussions/600#discussioncomment-6035224)).
And I have two notes:
First, as I also tested it in `llvm` modes I found two lines where you used `dr.cuda.TensorXf` which limits the use to `cuda` variants. In my case I changed these to use the Mitsuba type i.e. `mi.TensorXf`. I added comments [here](https://github.com/mitsuba-renderer/mitsuba3/pull/738/files#r1212294268) and [here](https://github.com/mitsuba-renderer/mitsuba3/pull/738/files#r1212294635).
Second, the problem described in the discussion #600 also seems to affect your implementation (only for `cuda` variants and with the `prb` integrator).
The problem was introduced with [this](https://github.com/mitsuba-renderer/drjit-core/compare/f657fe0c8c10f5d0aa60156905ffa3f760684b52...21bdbebfde8a7122f4d82e0a86df155bfb43ffb6#) change or the problem exists in that integrator and only became noticeable with the change. It seems to have something to do with `opaque` variables (they might need some documentation) and it might affect other projects that use the `prb` integrator.

**Comment by njroussel** [comment]
Hi @DoeringChristian 

Thanks for your review :+1:, I've been editing and improving the notebook/tutorial, and haven't made my way to the `largesteps.py` file yet. I should be done later today.

Regarding the flipped normals issue raised in #600, I've answered over there.

**Review by njroussel — APPROVED** [review]
Thank you everyone!

I'm happy with the current state of this PR/feature. We can revisit this in the future if we need to add support for textures or any other arbitrary structure actually.

**Pull request merged**